onnx/defs/tensor/defs.cc

```diff
@@ -89,7 +89,98 @@ ONNX_OPERATOR_SET_SCHEMA(
+          // iteration, but we record the Dimension. If targertShape dimension is 0, we attempt to infer the 
+          // dimension value. If it can be inferred, we set inferred dimension value; if not,
+          // we set the unresolvedZero flag. If targetShape dimension is positive,
+          // we set the dimension value in the outputShape. We track the product of 
+          // the dimensions we are setting outputShape in the outputProduct variable.
+          // The outputProduct will potentially be used for inferring a dimension marked -1.
+          auto* outputShape = ctx.getOutputType(0)->mutable_tensor_type()->mutable_shape();
+          TensorShapeProto::Dimension* negativeOneDim = nullptr;
+          const auto& dataInputTensorType = ctx.getInputType(0)->tensor_type();
+          bool unresolvedZero = false;
+          int64_t outputProduct = 1;
+          for (int i = 0; i < static_cast<int>(targetShape.size()); ++i)  {
+            // Add a new dimension to outputShape
+            auto* new_dim = outputShape->add_dim();    
+            if (targetShape[i] == -1) {
+              // Check if multiple -1's. If not, set negativeOneDim, marking this
+              // dimension to potentially be filled in later.
+              if (negativeOneDim) {
+                fail_shape_inference("Target shape may not have multiple -1 dimensions");
+              }
+              negativeOneDim = new_dim;
+            } else if (targetShape[i] == 0) {
+              // Check if data input has a shape, if the index i is within its bounds, and
+              // the dimension value is present at index i. If these conditions are satisfied,
+              // the dimension value can be inferred and set. Otherwise, set unresolvedZero flag.
+              if (dataInputTensorType.has_shape() && 
+                  i < dataInputTensorType.shape().dim_size() && 
+                  dataInputTensorType.shape().dim(i).has_dim_value()) {
+                const auto& dim_value = dataInputTensorType.shape().dim(i).dim_value();
+                new_dim->set_dim_value(dim_value);
+                outputProduct *= dim_value;
+              } else {
+                unresolvedZero = true;
+              }
+            } else if (targetShape[i] > 0) {
+              // Set the dimension value to targetShape[i]
+              new_dim->set_dim_value(targetShape[i]);
+              outputProduct *= targetShape[i];
+            } else {
+              // Check if value is less than -1; fail if so
+              fail_shape_inference("Target shape may not have dimensions less than -1");
+            }
+          }
+
+          // If negativeOneDim has been set, we attempt to infer its value. This can be done if 
+          // all other dimensions have been set (unresolvedZero is false) and complete shape
+          // information is present for the data input tensor.
+          if (negativeOneDim && !unresolvedZero) {
```

Comment:
I think we can still infer the negativeOneDim when there is unresolvedZero, e.g.
```
input shape: 1 2 a 3
target shape: 1 1 0 -1
=>
output shape: 1 1 a 6
```

Reply:
True. To infer in this case, we need to keep track of which indices are unresolvedZero's. Only when the data input tensor has missing values corresponding to unresolvedZero's can we infer